File: data structures and algorithms/java/06_073.java
import java.util.ArrayList;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.List;

public static List<Integer> kSort(int[] arr, int k) {
   int n = arr.length;
   List<Integer> result = new ArrayList<>();
   
   // Create buckets for the range of possible values
   Deque<Integer>[] buckets = new ArrayDeque[k+1];
   for (int i = 0; i <= k; i++) {
       buckets[i] = new ArrayDeque<>();
   }
   
   // Process elements in order
   for (int i = 0; i < n; i++) {
       // Place elements in appropriate buckets
       int bucketIdx = Math.min(k, i);
       buckets[bucketIdx].addLast(arr[i]);
       
       // If we have filled k+1 buckets, start extracting
       if (i >= k) {
           int minBucket = getMinBucketIndex(buckets);
           result.add(buckets[minBucket].removeFirst());
       }
   }
   
   // Extract remaining elements
   while (anyBucketsNotEmpty(buckets)) {
       int minBucket = getMinBucketIndex(buckets);
       result.add(buckets[minBucket].removeFirst());
   }
   
   return result;
}

private static int getMinBucketIndex(Deque<Integer>[] buckets) {
   int minValue = Integer.MAX_VALUE;
   int minBucket = 0;
   
   for (int i = 0; i < buckets.length; i++) {
       if (!buckets[i].isEmpty() && 
           buckets[i].peekFirst() < minValue) {
           minValue = buckets[i].peekFirst();
           minBucket = i;
       }
   }
   
   return minBucket;
}

private static boolean anyBucketsNotEmpty(
       Deque<Integer>[] buckets) {
   for (Deque<Integer> bucket : buckets) {
       if (!bucket.isEmpty()) {
           return true;
       }
   }
   return false;
}

File: data structures and algorithms/cs/06_073.cs
using System;
using System.Collections.Generic;

public static List<int> KSort(int[] arr, int k) {
   int n = arr.Length;
   List<int> result = new List<int>();
   
   // Create buckets for the range of possible values
   Queue<int>[] buckets = new Queue<int>[k+1];
   for (int i = 0; i <= k; i++) {
       buckets[i] = new Queue<int>();
   }
   
   // Process elements in order
   for (int i = 0; i < n; i++) {
       // Place elements in appropriate buckets
       int bucketIdx = Math.Min(k, i);
       buckets[bucketIdx].Enqueue(arr[i]);
       
       // If we have filled k+1 buckets, start extracting
       if (i >= k) {
           int minBucket = GetMinBucketIndex(buckets);
           result.Add(buckets[minBucket].Dequeue());
       }
   }
   
   // Extract remaining elements
   while (AnyBucketsNotEmpty(buckets)) {
       int minBucket = GetMinBucketIndex(buckets);
       result.Add(buckets[minBucket].Dequeue());
   }
   
   return result;
}

private static int GetMinBucketIndex(Queue<int>[] buckets) {
   int minValue = int.MaxValue;
   int minBucket = 0;
   
   for (int i = 0; i < buckets.Length; i++) {
       if (buckets[i].Count > 0 && 
           buckets[i].Peek() < minValue) {
           minValue = buckets[i].Peek();
           minBucket = i;
       }
   }
   
   return minBucket;
}

private static bool AnyBucketsNotEmpty(Queue<int>[] buckets) {
   foreach (Queue<int> bucket in buckets) {
       if (bucket.Count > 0) {
           return true;
       }
   }
   return false;
}